sql/sql_parse.cc

```diff
@@ -9048,15 +9050,25 @@ kill_one_thread(THD *thd, longlong id, killed_state kill_signal, killed_type typ
 
     if (((thd->security_ctx->master_access & SUPER_ACL) ||
         thd->security_ctx->user_matches(tmp->security_ctx)) &&
-	!wsrep_thd_is_BF(tmp, false))
+        !wsrep_thd_is_BF(tmp, false))
     {
-      tmp->awake_no_mutex(kill_signal);
+      /* killing a thread in a cluster node, must happen in controlled way to
+         avoid mutex locking ordering violation. Here we test that victim is 
+         not high priority, and use wsrep_abort_thd() for killing the victim */
+      if (WSREP(thd) && !wsrep_thd_is_BF(tmp, true))
+      {
+        mysql_mutex_unlock(&tmp->LOCK_thd_kill);
+        wsrep_released_LOCK_thd_kill= true;
```

Comment:
This is the execution of manual KILL command, and it uses find_thread_by_id() to locate the victim and  locking THD::LOCK_thd_kill of the victim. However, after that the LOCK_thd_kill is released and wsrep_abort_thd() is called with no mutexes locked. wsrep_abort_thd() executes same BF aborting call path, that we normally do for MDL conflict resolving, and it will eventually dive into innodb side and executes wsrep_innobase_kill_one_trx(), where it merges with the BF aboring code path, which originates from innodb lock manager